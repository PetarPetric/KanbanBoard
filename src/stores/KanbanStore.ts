import { nextTick, ref } from 'vue';
import { defineStore } from 'pinia';
import type { IColumn } from "../types/Column"
import type { Toast } from "../types/Toast";
import type { ITask } from "../types/Task";
import Repository from "@/services/apiService/APIFactory";
const ColumnRepository = Repository.get("columns");
const TasksRepository = Repository.get("tasks");

export const useKanbanStore = defineStore('KanbanStore', () => {
  const listOfColumns = ref<IColumn[]>([]);
  const toast = ref<null | Toast>(null);
  const columnUpdated = ref(null);

  // Call toast component
  const setToast = (type: string, text: string) => {
    toast.value = {
      type,
      text
    };

    setTimeout(() => {
      toast.value = null;
    }, 1500);
  }

  // Fetch columns
  const fetchColumns = async () => {
    const { data } = await ColumnRepository.get()
    listOfColumns.value = data.sort(
      (a: IColumn, b: IColumn) => a.columnPosition - b.columnPosition
    )
  }

  // Add column api call 
  const addColumn = async (name: string, columnPosition: number) => {
    const form: IColumn = {
      title: name,
      boardId: 0,
      columnPosition,
    }
    await ColumnRepository.create(form);
    await fetchColumns();
    setToast('success', 'Successfully added column');
  }

  // Remove column
  const removeColumn = async (id?: number) => {
    await ColumnRepository.delete(id)
    await fetchColumns();
    setToast('success', 'Successfully removed column');
  }

  // rearrange columns when doing drag and drop like in jira
  const reArrangeColumns = async (id: number, columnPosition: number) => {
    const form = {
      columnPosition
    }
    await ColumnRepository.patch(form, id);
    setToast('info', 'Successfully rearranged columns');
  }

  // Edit columns
  const editColumn = async (id: number, name: string, columnPosition: number) => {
    const form: IColumn = {
      title: name,
      boardId: 0,
      columnPosition,
      id
    }
    await ColumnRepository.update(form, id);
    await fetchColumns();
    setToast('info', 'Edited column');
  }

  // Add task api call
  const addTask = async (columnId: number, title: string, taskPosition: number) => {
    const form: ITask = {
      title,
      description: '',
      columnId,
      taskPosition,
    }
    await TasksRepository.create(form);
    setToast('success', 'Successfully added task');
  }

  // Fetch tasks
  const fetchTasks = async (columnId: number) => {
    const { data } = await TasksRepository.get(columnId)
    return data;
  }

  const fetchSingleTask = async (id: number) => {
    const { data } = await TasksRepository.getSingle(id)
    return data;
  }

  // Move task actions
  const moveTask = async (id: number, columnId?: number, taskPosition?: number) => {
    const form = {
      columnId,
      taskPosition
    }
    const { data } = await TasksRepository.patch(form, id);
    setToast('info', 'Successfully moved task');
    return data;
  }

  // Remove task
  const removeTask = async (id: number) => {
    await TasksRepository.delete(id)
    await fetchColumns();
    setToast('success', 'Successfully removed a task')
  }

  // Edit task
  const editTask = async (form: ITask) => {
    const { data } = await TasksRepository.update(form);
    setToast('info', 'Edited Task');
    return data;
  }

  return {
    fetchColumns,
    addColumn,
    removeColumn,
    editColumn,
    reArrangeColumns,
    addTask,
    fetchTasks,
    fetchSingleTask,
    removeTask,
    moveTask,
    editTask,
    setToast,
    toast,
    listOfColumns,
    columnUpdated
  }
});