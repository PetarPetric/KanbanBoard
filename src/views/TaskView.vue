<script setup lang="ts">
import { onMounted, ref } from "vue";
import { useRouter, useRoute } from "vue-router";
import { useKanbanStore } from "@/stores/KanbanStore";
import ModalComponent from "@/components/ModalComponent.vue";
import type { ITask } from "@/types/Task";
const $route = useRoute();
const $router = useRouter();
const store = useKanbanStore();
const taskData = ref<ITask | null>(null);

const closeModal = () => {
  $router.push("/");
};

const getTaskData = async () => {
  let taskId;
  if ($route?.params?.id) {
    taskId = Number($route.params.id);
  }
  taskData.value = await store.fetchSingleTask(taskId as number);
};

const updateTask = async () => {
  await store.editTask(taskData.value as ITask);
  $router.push("/");
};
onMounted(async () => {
  await getTaskData();
});
</script>

<template>
  <ModalComponent @close-modal="closeModal">
    <template v-slot:body>
      <div v-if="taskData" class="task-details">
        <label for="task-title">Task Title</label>
        <input
          v-if="taskData.title"
          id="task-title"
          placeholder="Task Title"
          v-model="taskData.title"
        />
        <label for="task-description">Task description</label>
        <input
          id="task-description"
          placeholder="Task Description"
          v-model="taskData.description"
        />
        <label for="columns">Task Column</label>
        <select v-model="taskData.columnId" name="columns" id="columns">
          <option
            v-for="column in store.listOfColumns"
            :value="column.id"
            :key="column.id"
          >
            {{ column.title }}
          </option>
        </select>
      </div>
    </template>
    <template v-slot:footer>
      <div class="task-details__footer">
        <button @click="closeModal" class="btn btn--cancel">Cancel</button>
        <button @click="updateTask" class="btn btn--save">Save</button>
      </div>
    </template>
  </ModalComponent>
</template>

<style scoped lang="scss">
@import "@/styles/taskView.scss";
</style>
