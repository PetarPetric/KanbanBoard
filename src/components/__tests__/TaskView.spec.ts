import { mount, VueWrapper, flushPromises } from "@vue/test-utils";
import { createRouterMock, injectRouterMock } from 'vue-router-mock'
import { createTestingPinia } from '@pinia/testing'
import { useKanbanStore } from "@/stores/KanbanStore";
import TaskView from "@/views/TaskView.vue";

describe("task view component", () => {
  let wrapper: VueWrapper<any>;
  let store: any;
  const router = createRouterMock()

  beforeEach(() => {
    injectRouterMock(router);
    wrapper = mount(TaskView, {
      global: {
        plugins: [createTestingPinia({
          createSpy: vi.fn
        })],
      }
    });
    store = useKanbanStore();
  });

  afterEach(async () => {
    wrapper.unmount()
  })

  it('calls correct api', async () => {
    await wrapper.vm.$router.push({ path: '/task/3' })
    expect(wrapper.vm.$route.path).toBe(`/task/3`);
    expect(store.fetchSingleTask).toHaveBeenCalledTimes(1);
    expect(store.fetchSingleTask).toHaveBeenLastCalledWith(wrapper.vm.$route.params.id);
  })

  it('should render correctly', async () => {
    await wrapper.vm.$router.push({ path: '/task/3' })
    expect(wrapper.html()).toMatchSnapshot();
  })
});
