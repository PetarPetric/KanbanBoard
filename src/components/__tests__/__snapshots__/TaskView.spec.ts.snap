// Vitest Snapshot v1

exports[`task view component > should render correctly 1`] = `
"<div class=\\"modal__wrapper\\" data-v-6280910c=\\"\\" data-v-47886999=\\"\\">
  <div class=\\"modal__container\\" data-v-6280910c=\\"\\"><button class=\\"modal__btn-close\\" data-v-6280910c=\\"\\"><svg width=\\"1em\\" height=\\"1em\\" viewBox=\\"0 0 16 16\\" fill=\\"currentColor\\" role=\\"img\\" focusable=\\"false\\" data-v-6280910c=\\"\\">
        <path fill-rule=\\"evenodd\\" d=\\"M13.854 2.146a.5.5 0 0 1 0 .708l-11 11a.5.5 0 0 1-.708-.708l11-11a.5.5 0 0 1 .708 0Z\\"></path>
        <path fill-rule=\\"evenodd\\" d=\\"M2.146 2.146a.5.5 0 0 0 0 .708l11 11a.5.5 0 0 0 .708-.708l-11-11a.5.5 0 0 0-.708 0Z\\"></path>
      </svg></button>
    <div class=\\"modal__body\\" data-v-6280910c=\\"\\"></div>
    <div class=\\"modal__footer\\" data-v-6280910c=\\"\\">
      <div class=\\"task-details__footer\\" data-v-47886999=\\"\\"><button class=\\"btn btn--cancel\\" data-v-47886999=\\"\\">Cancel</button><button class=\\"btn btn--save\\" data-v-47886999=\\"\\">Save</button></div>
    </div>
  </div>
</div>"
`;
